### ISSUE TITLE
Provide output for --fix'ed code

### ISSUE THREAD

**Issue by jdb8** [eslint/eslint#3942]
**What version are you using?**
v1.5.1

**What did you do?**
Ran --fix on files which had fixes to be applied, containing no unfixable errors.

**What happened?**
eslint returned 0, a success error code.

**What did you expect to happen?**
Since errors were detected and fixed, either an error code of 1 or some non-zero error code would differentiate a successful 'no linting errors detected' to an 'unsuccessful' run.

This is a tricky issue because an argument can be made that the fixer succeeding constitutes a successful execution. However, the use-case which causes issues is one involving pre-commit hooks which require commits to fail if fixes have been made. It is often necessary to make wrapper tools aware of changes made to the file system.

My proposal, assuming there is no agreement to change the exit code itself, is to provide output for the number of fixes made when calling CLIEngine.outputFixes. Right now when creating a report with `fix: true` set, all errorCounts are 0 despite errors being detected and fixed. Including the output at the outputFixes step would allow wrappers to at least fix the code themselves and output the error codes they expect in this case.

If I've missed an existing way to check for fixes being made (ideally along with filenames of fixed files) please let me know. The only way I can currently think of doing this is to run eslint twice, once without `--fix` and once with, and compare the errorCount for both runs. This doesn't seem ideal.

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
> However, the use-case which causes issues is one involving pre-commit hooks which require commits to fail if fixes have been made

Then don't configure it to make fixes in your pre-commit hook. Have it check, and upon failing give the user the option to automatically make the fix.

:-1:. When the fixing succeeds, exit code should definitely be zero.

**Comment by jdb8** [comment]
@michaelficarra My main proposal is to give scripts which use eslint's CLIEngine to fix code the information on the files which were fixed. I'm not pushing to change the error code output of eslint as a whole (despite my personal feelings on the matter), but I wanted to provide context for my request.

Your suggestion wouldn't work for me since I'm trying to integrate eslint's fixer with https://github.com/pre-commit/pre-commit, which isn't designed to behave interactively. For other fixers, such as fixmyjs or autopep8, I can either use them directly (with fixmyjs since an error code of 1 is returned on fixed files) or write wrapper code to handle fixes (with autopep8, which exposes a list of fixed files). Unfortunately for eslint it seems right now I can do neither, hence this issue.

**Comment by nzakas** [comment]
@jdb what information are you looking for? Number of fixes applied? Just true/false that fixes were applied?

**Comment by jdb8** [comment]
@nzakas Any of the above would be enough, in order to provide a non-zero error code for pre-commit. Ideally though, an array of filenames would be great in order to provide a nicer console output (in the same way as https://github.com/pre-commit/pre-commit-hooks/blob/master/pre_commit_hooks/autopep8_wrapper.py#L20).

**Comment by IanVS** [comment]
if you are using CLIEngine, couldn't you examine the reports which are returned, and check them for `fix` before running [outputFixes()](http://eslint.org/docs/developer-guide/nodejs-api#outputfixes)?

**Comment by nzakas** [comment]
@IanVS not all fixes are applied. 

@jdb8 the array of filenames that what?

**Comment by jdb8** [comment]
@IanVS that seems like exactly what I want, but indeed if not all fixes are applied it won't work for me. 

@nzakas the array of filenames that had fixes applied during `--fix` or `eslint.CLIEngine.outputFixes(report)` if running from a script.

I'm curious what you mean about not all fixes being applied though. Looking at `outputFixes`, it seems to simply check if an `output` key exists, and if so, write to the file with said output: https://github.com/eslint/eslint/blob/master/lib/cli-engine.js#L447

Therefore using 
```
var report = engine.executeOnFiles(filenames);
var filesWithFixes = report.results.filter(function(result) {
    return result.hasOwnProperty('output');
});
```
in my wrapper (as per @IanVS's suggestion) seems to do the same thing, and provides me with the filenames. Is there a case where this wouldn't work?

It still would be nice to have a return value from `outputFixes` itself which provides this info (or even a command-line argument to provide this to avoid the need for any wrapper script at all), but I'm happy for now regardless assuming this code works.

**Comment by nzakas** [comment]
Yes, that will do the same thing. `applyFixes()` basically just loops through and copies `output` into `filePath` from the results.

What you're seeing returned from `executeFiles()` is files that have had some fixes applied, however, not all fixes that were indicated in the results were applied to the file. So, you if you had 10 possible fixes, you might end up with only 1 applied due to overlapping ranges, for example.

What @ianvs was referring to was the `fix` tag that is applied to each message that has an appropriate fix. But as I said, not all of these fixes will be applied. The `output` property is added to each file's result only if at least one fix has been made.

Given that you can determine which files will be output in the way you mentioned above, I'm unsure of the usefulness of returning that list from `applyFixes()`. It doesn't seem like a natural return value for that function, whereas I could more easily see an argument for returning `true` if it wrote any files or `false` if not.

**Comment by IanVS** [comment]
My mistake, I haven't worked with autofix very much, and was just looking at the docs and saw the `fix` property on the messages.   I wasn't aware of `output`, which seems much better to use.

So, @jdb8 it sounds like you have something that will meet your need, yes?

**Comment by jdb8** [comment]
@IanVS yep, the code I pasted above should suit my needs. Thanks!

@nzakas I think a `true`/`false` return value for `outputFixes` would make sense, although I won't be using it in this case necessarily. At this point given the functionality for accessing fixed filenames is accessible via the CLIEngine method, I don't feel strongly that it should be added anywhere else. I'm sure if someone has a need for alternative behaviour they can make a new proposal.
